### PR TITLE
build: use esbuild plugin

### PR DESCRIPTION
**Pull request by aulneau**
Hi there!

I made a [PR for jotai](https://github.com/pmndrs/jotai/pull/378) a few days ago updating the build tooling to use esbuild as the typescript transpiler and @dai-shi requested I make PRs for [zustand](https://github.com/pmndrs/zustand/pull/343) and valtio, so here we are! The main reason for this change is because of how much faster esbuild is at compiling javascript from typescript:

Before
✨  Done in 24.54s.

After
✨  Done in 9.38s.

There are also some very minor bundle size changes with this update.

### General overview of changes
- added `esbuild` and `rollup-plugin-esbuild`
- replaced the `typescript` plugin in `createESMConfig` and `createCommonJSConfig` to use esbuild
- added `id.includes('@babel/runtime')` to the externals check [per this recommendation](https://github.com/rollup/plugins/tree/master/packages/babel#babelhelpers)
- updated the node version used in the codesandbox ci to 12 (due to esbuild requirement)
- added `tslib` as a dev dependency due to it being missing for the typescript plugin

### CODE REVIEW

**Comment by dai-shi** [comment]
Hi, thanks for your work!
Would you also apply the same rollup config and concurrently technique in jotai here?

Two reasons:
- that would speed up even more
- we want to have the same structure for all three libs for better maintainability (which wasn't the case before.)

**Comment by aulneau** [comment]
> Two reasons:
>
> - that would speed up even more
> - we want to have the same structure for all three libs for better maintainability (which wasn't the case before.)

Done! I actually think the speed element is the same, but I agree, it's better to have consistent structure between each.

**Comment by thelinuxlich** [comment]
I hope that esbuild can replace babel macro in the future :)

File: rollup.config.js [hunk 1]
```diff
@@ -47,21 +59,25 @@ function createCommonJSConfig(input, output) {
     external,
     plugins: [
       resolve({ extensions }),
-      typescript(),
+      getEsbuild('es2015'),
```
Comment (dai-shi) [inline rollup.config.js:62]:
```suggestion
```

according to https://github.com/pmndrs/zustand/pull/343#issuecomment-808745679 @aulneau

Comment (aulneau) [reply rollup.config.js]: done!

File: babel.config.js [hunk 2]
```diff
@@ -16,6 +16,7 @@ module.exports = (api, targets) => {
       ],
     ],
     plugins: [
+      '@babel/plugin-transform-runtime',
```
Comment (dai-shi) [inline babel.config.js:19]: we don't need this?

Comment (aulneau) [reply babel.config.js:19]: I might be wrong, but I think we do -> https://github.com/pmndrs/jotai/pull/378#discussion_r602957217

I set this as I wanted to get rid of the warning when bundling.

**Review by dai-shi — APPROVED** [review]
LGTM

**Pull request merged**